### DISCUSSION TITLE
Confirmation of the reason for the two-step constraint-based installation process

### DISCUSSION THREAD

**Discussion by michalc** [apache/airflow#35743] [Q&A] [answered]
After watching https://www.youtube.com/watch?v=zPjIQjjjyHI and reading https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html I'm still wrestling a bit with the reason for the two-step and constraint-based installation process, which is (repeating here so I can be corrected if I've misunderstood)

1. Install Airflow with extras with the constraint file that pins to _exact_ versions
2. Install other dependencies _without_ the constraint file, which loosens Airflow + extras versions to a range specified in setup.cfg/setup.py

Is it essentially to make it so "dependencies are allowed to differ from the versions in Airflow's constraint file, but _only_ if necessary by the install of the second step"

In other words, it keeps the flexibility of just using the ranges of setup.cfg/.py, but reduces risk of unexpected incompatibilities due to new versions of packages released that haven't been directly tested with Airflow - they only happen if necessary by the second step. The second step won't just upgrade everything just because it can, it will upgrade the _minimal_ set of packages to complete the install.

In a way it works around a limitation of Python dependency ranges in setup.py/cfg - there isn't a way to say "I really want this specific version, but this wider range of versions is fine if something else needs it"

Is this right?

Context: we have a requirements.in/requirements.txt-based install process, with what is essentially a single `pip install -r requirements.txt` install when deployed to production, and I'm not sure if we can change that in the short term. So I want to understand this to see if I can at least get close to Airflow's recommended install mechanism.

**Comment by potiuk** [comment, marked as answer]
> In a way it works around a limitation of Python dependency ranges in
setup.py/cfg - there isn't a way to say "I really want this specific
version, but this wider range of versions is fine if something else needs
it"

Yes. You got it right. But it's not a "working around" limitation. There is
simply no dependency system in the world that can solve it because this is
split responsibility. First part of it is "maintainer's responsibility" and
second is "user responsibility". So - out of necessity - this has to be a 2
step process. Where 1) is something that maintainers do where 2) is
something that the user needs to take a reign (and responsibility) of.

It's basically to cover dualism of airflow being application (where
maintainers want to pin dependencies) and library (where users want to have
freedom and leave them open-ended). Airflow is both. You want to
reproducibly install it and let it drive dependencies it needs (as an
application) but as DAG developer you want to treat airflow as a
library because you might want to upgrade to another version of dependency
that Airflow has.

It is mostly coming together with airflow dependencies being very rarely
upper-bound - and more of than not constraint files containing "newest"
versions of those (if they are not conflicting with other requirements).
This means that we want to let users freely upgrade to newer versions of
those as they are released AFTER we released a particular version of
Airflow. This is the main use case. Probably 300 of our 670 dependencies in
a year from now will have newer versions and our users often want to
upgrade them - without having to upgrade Airflow itself.

Example: say you have airflow 2.7.3 with constraints for python 3.10
https://github.com/apache/airflow/blob/constraints-2.7.3/constraints-3.10.txt
- there you have "identify==2.5.31"
But yesterday - identify released version 2.5.32
https://pypi.org/project/identify/#history - and if for whatever reason you
need that newer version to be used when you develop DAG - you can do it
because we do not upper-bind it. But - since we released 2.7.3 already 3
weeks ago, we had no idea (because we have no crystal ball) whether
identify will release a new version and whether it will be still working
with Airflow. You might try to rely on various "breaking/non-breaking"
signals - maybe we could limit identify to < 2.6 or <3.0 - but this is very
wrong. There is no "universal" notion that 2.6.0 or 3.0 will be breaking
for us where 2.5.32 will not be. It might go both directions - maybe 2.5.32
will be breaking (happened multiple times with multiple dependencies) or
that 3.0.0 will still be working fine. We simply have no crystal ball to
tell it for 670 dependencies of ours.

We do however limit some important dependencies that we know what scheme
they use and where we know/expect/foresee it will be breaking - see our
setup.cfg where we have few of those and meticulously document WHY we are
limiting them.

Few examples:

    # Alembic is important to handle our migrations in predictable and
performant way. It is developed
    # together with SQLAlchemy. Our experience with Alembic is that it very
stable in minor version
    alembic>=1.6.3, <2.0

    # Flask 2.3 is scheduled to introduce a number of deprecation removals
- some of them might be breaking
    # for our dependencies - notably `_app_ctx_stack` and
`_request_ctx_stack` removals.
    # We should remove the limitation after 2.3 is released and our
dependencies are updated to handle it
    flask>=2.2,<2.3


So 1) helps maintainers to come up with installation with "constraints"
being reproducible installation with "known working versions" - so that
users can rely on them working. Where 2) allows our users to (mostly)
upgrade dependencies to newer ones.

There is also another use case - It also allows you to downgrade
dependencies because there might be some cases where you want to use an
older version of a library (older than our constraints) - for example
because your custom code does not work with it. For that we have
"min-requirements" - which I hope soon will also be quite well reflecting
the "min" limits that airflow should work with  - I have an opened issue
for that https://github.com/apache/airflow/issues/35549 - currently it's
"mostly" right but might be that we need to bump some older dependencies.

Now - how your upgrade process should look like (this is really what you
need to know): - the solution is simple (and it is actually explained in my
presentation) = use/generate your own constraints (and turn them into
requirements). And it's - in principle - rather simple:

1) Whenever you update Airflow (say to 2.7.4), start with
Airflow's constraints and install it in reproducible way with "our"
constraints

2) Update whatever you need to upgrade - you need to know what specific
expectations you have in your environment, which libraries you need to
downgrade or upgrade. Apply those, solve conflicts if needed. It's always
YOUR responsibility to keep track and be aware what specific expectations
your installation has - we have no idea to know what they are in all the
10s of thousands cases of our users - you are the only one who can know it
and keep track of it (and update - because sometimes you might need to
update those if they are conflicting with new airflow and fix your code).

3) Test if it works after your upgrade.

4) `pip freeze > your_constraints.txt` -> and use those as your
requirements -> that should become your "master" set of requirements (they
will be your requirements really at this stage because you will require
them and use the file with `-r`.

Note that the moment you do 1) - install airflow reproducibly - this is
where we can "guarantee" (reproducible installation) ends. Then all the
burden is on you. We cannot solve all the conflicts and dependency
variations (matrix of 670 of them) for everyone, so what we do is we tell
you "And now, if you want to diverge, you can, but you need to make sure it
works". In other words - with great powers come great responsibilities. So
when you want to diverge -  it's your responsibility as a
deployment manager to track whatever special needs you have, and to make
sure that it still works. Because (for example if identify 2.5.32 would be
breaking - we had no idea about it when we released 2.7.3 because we had no
crystal ball). Sometimes you might also need to fix your code if conflicts
are impossible to solve. It might be that for whatever reason new Airflow
has bumped "min-requirement" for a library and that your code needs a lower
version of it, And in this case what you really need to do is to make sure
your code is made compatible with the new version of the library that
Airflow bumped. And again - it's on you to solve - we simply cannot do it
for you.

J.

**Reply by michalc** [reply]
> >  In a way it works around a limitation of Python dependency ranges in
setup.py/cfg - there isn't a way to say "I really want this specific
version, but this wider range of versions is fine if something else needs
it"

> Yes. You got it right

Thanks! So I _think_ I understand

> But it's not a "working around" limitation. There is simply no dependency system in the world that can solve it because this is split responsibility

But I maybe disagree with this? Right now the supported way is to use pip twice, and in both steps Airflow-determined version "restrictions" are used. It doesn't seem beyond the realm of possibility that that these be combined somehow. Maybe not via some sort of extended range definition now I think about it, but say if the specification of that allowed some sort of `[project.preferred-constraints]` section in one of Airflow's config files. Then also any sort of pip freeze (or pip-compile?) output could also explain why something is the version it is, including why any preferred-constraint were violated. It would help make such output reviewable as well: "pay attention to this - it is in all allowed ranges, but it violates preferred-constraints in X because of Y"

I guess I think achieving such a thing is probably a project, involving many people across multiple projects, so I'm not really "requesting" this, but just sort of wondering what the absolute ideal is. If you could just click your fingers and magically change stuff, would the above be a better state of affairs? Or is the two step process really the ideal?

**Reply by potiuk** [reply]
I think It's possible to combine those two steps in a single commante, it would still be two steps installation however, where first part is "maintainer" provided and the second is something that "user" is responsible for. 

There is no way around this separation - it's logically impossible to bypass that because you have different people at different point of lifecycle of the package making different decisions (maintainers need to decide at release time what is the "golden" set of packages, and users might want to decide at installation time what is the deviation from that). The second decision will be different for different users, and it will change over time where the first one is common for everyone and decision is made at the release time.

Yes, you can "hide it" and provide standard tooling for that and for example have a single `pip` command to do both. But this will be just a syntactic sugar - rather than having two commands you would have two of them doing two steps and effectively handing over the responsibility. 

And yes I thought about making it into a more standard python tooling. But this is a process that is not guaranteed to succeed and requires a lot of time, coordination and most-of-all agreement in Python Packaging group (and there are people who are very opinionated on what should get accepted or not, so there is no guarantee this process will succeed  -because while sound, this is really only needed for projects like airflow - with 100s of dependencies and specific requirements to be both application and library. From Python Packaging point of view - this is an edge case (and I agree it is).

But yes we could implement it and I have a vision in my head how it could look like:

For that you need two things:

1) Define a standard (you will need to to follow PEP process for it) on how to store this kind of "golden" set o dependencies. Sort of "remote lock file". But currrently Python Packaging cannot even agree on the need of standard for "development lock file" - each o of the packaging tools uses their own "standard" there and the community is divided on how it should look like and whether it should be standardised in the first place. I decided to wait until this discussion settles and some PEPs are not only approved but also implemented by some tools for the "local lock" because then we could think of the next step - sharing the lock remotely for users installing your package (as constraints).

2) Then - you would have to implement that "remote lock" file support in the packaging tools you can try to make it part of one of the tools (`pip`, `poetry`, `rey` recently and probably dosen of othet tools (and if we make point 1) as widely approved PEP - it has a chance that maintainers of those tools would be happy to accept it.

Or there is another option: you can develop your own packaging tool. Which leads us quickly to this situation:

![image](https://github.com/apache/airflow/assets/595491/0eb0a2a6-d375-4b97-81d9-6fd7ae9a1364)

Which, I think taking into consideration of the current packaging tools environment in Python is probably a very bad idea.


So, summarizing - yes, you could make the approach we use in Airflow part of the standard approach (still it will only be a syntactic sugar merging two commands into one), even though it is a bit of an edge case, but it will take years and a lot of people to agree on the direction.

Which would be great to do, but considering the effort and amount of communication needed, might happen in a few years but I would only approach doing it once some of the standards are settled there.

In the meantime, having two commands and our own "approach" to share the constraints in GitHub  seems like something you can use now rather than wait several years for standards and tooling to catch up

**Reply by michalc** [reply]
Understood! Thank you for your detailed replies - I think I mostly understand the situation

> There is no way around this separation - it's logically impossible to bypass that because you have different people at different point of lifecycle of the package making different decisions (maintainers need to decide at release time what is the "golden" set of packages, and users might want to decide at installation time what is the deviation from that)

To respond to this - even ignoring Airflow, this situation already exists I think. Maintainers of each package set a range of compatible dependencies, where implicitly the "golden" version of each is the "highest version available in the range". Then users decide which specific versions to install. The golden ones always get installed, unless some other dependency that the user is trying to install demands something else. And this is all in the one step using pip (or poetry or whatever...)

It looks to me that Airflow just customises this(?) - its golden ones are not necessarily the highest versions available in the supported range.

(It is touch more complicated because I think Airflow's constraint file sets golden versions of all the transitive dependencies, including for all its extras, but I'm more arguing that concept of "two different lifecycle steps applied at once" is already there really, as well as a sort of implicit "golden" dependency version)

> In the meantime, having two commands and our own "approach" to share the constraints in GitHub seems like something you can use now rather than wait several years for standards and tooling to catch up

And oh yes of course!

**Reply by michalc** [reply]
> this is really only needed for projects like airflow - with 100s of dependencies and specific requirements to be both application and library. From Python Packaging point of view - this is an edge case (and I agree it is)

There is something else... I'm starting to think this is not an edge case, and that Airflow being both an application and a library is a bit of a red herring.

All packages that have dependencies, unless they are really trivial in number, can only test with a very small combination of them. Often just one of each even. This is their "golden" set of dependencies* (or rather - it would be if there were some way to communicate that downstream)

Sure - Airflow does have loads of dependencies, but the concept of having tested only a few of them while wanting users to have the flexibility of going beyond those versions tested, I am now thinking isn't an "application and library" thing. It's more an "any package with dependencies" thing.

(*There could be multiple golden sets in some cases if the package maintainers test with multiple versions of dependencies - I do in some of my packages for example)

**Reply by potiuk** [reply]
> All packages that have dependencies, unless they are really trivial in number, can only test with a very small combination of them. Often just one of each even. This is their "golden" set of dependencies* (or rather - it would be if there were some way to communicate that downstream)

I quite disagree. I think our case is pretty special and I would be very far from telling others to do the same or similar.

The big difference of Airflow is that it has two different sets of users:

1) Deployment Managers -> who treat airflow as application to install. They just want the application to "work". The problem of those apps is solved mostly by having good development tooling that allow you to update the dependencies while you develop things. If your elease your application often, applying security patches and upgrading dependencies, I think application users to not need to care about dependency problem too much. There your really have only the 1st part of the problem (make sure that maintainers solve it at release time).

2) DAG authors -> who author DAGs. And they are programmers who want to write their own Python code that uses their own libaries. This is where airflow dualism becomes a problem. It's not a red-herring. We have a LOT of users who deal with those problems on a daily basis. For example MWAA has special, dedicated solution (which is not perfect and I told them so) how to manage the dependencies for those cases. And there multiple problems with it.

If you "only" have application - `pipx` and `virtualenv` largely solve the problem - because each application can have own set of dependencies. 

But we have both.

But , there is also one more avenue we can take (and with our multi-tenancy effort which I am leading it might become a viable solution in the future. We can SPLIT airflow into two pieces:

* UI + Scheduler -> with its own dependencies (library)
* Parsing and executing DAGs -> where each DAG can even have their own, completely different set of dependencies, and they might not even ave the "full" airflow set of dependencies.

And it's quite likely that in the near future we move there (this is what I am preparing for quite some time now). We could separate out `airflow-runtime` = a minimum library that will allow to parse and execute tasks and "airflow-core" - whicih will be UI  and scheduler (or maybe even separate `airflo-ui` and `airflow-scheduler`) . This has it's own set of challenges - because while it will simplify dependency management, it will complicate deployment, and we cannot make it the "default" settings for all users (airfllow deployment is sufficiently complicated already for them). Plus it will awfully complicate the way airflow development is done (but solutions to that will be implemented sooner rather than later - it has already started with some recent packaging changes that will be continued).

But this will solve the dualism we have now pretty completely. I think in ~1.5 years if everything goes smoothly, we might actually get there. But we need to move slow to get there and for me this is the ultimate way how the problem of dependencies will be solved. 

At least this is my vision how I would like the dependency problem to be solved for Airflow.

**Reply by michalc** [reply]
> The big difference of Airflow is that it has two different sets of users:
> Deployment Managers  (...) -> who treat airflow as application to install
> DAG authors (...)

I guess my world/thinking is different somehow... because of the existence of DAG authors, Airflow is no longer just an application at all. It's now more like a framework, a bit like Flask say, which you use to make your own custom application. Just like you might use Flask to make an application with have loads of endpoints, you might use Airflow to make an application with loads of DAGs. You might have loads of extra dependencies (or specific versions of dependencies) in a Flask app to write those endpoints, just like in Airflow you have dependencies for specific DAGs. I wouldn't say Flask has this sort of dualism, so to me, neither has Airflow.

I might have to ponder though - I usually need a lot of time to mull things over. Usually in the shower or walking the dog :-) 

> We could separate out airflow-runtime = a minimum library that will allow to parse and execute tasks and "airflow-core" - whicih will be UI and scheduler (or maybe even separate airflo-ui and airflow-scheduler) 

But this sounds great to me. I think it would make Airflow similar to Dagster? It did feel good when we experimented with Dagster for this reason I think - the split "felt right". Our main Airflow setup is closed, but our experimentation with Dagster is open at https://github.com/uktrade/data-flow-2. The setup is essentially a basic version of the setup we use for our custom web applications, hence me maybe struggling to see the dualism.

**Reply by michalc** [reply]
I have just realised something. When I describe our Airflow setup, I say it's

> an instance of [Apache Airflow](https://airflow.apache.org/) together with the code of pipelines that transfer and transform data

... so do I sort of hint at the "application" part. But I think I do it deliberately to make it an easier entry point to the project for DAG authors. They shouldn't focus too much on the Airflow part, or what a framework is. They _should_ think of it just like an application that runs their DAGs, at least at the beginning.

But when I think about what our setup "really" is, it's more like a custom web application that's built on a framework. Others are built on say Flask, Django, FastAPI... this one is built on Airflow (I know Airflow itself is built on Flask... but I think it's perfectly possible for one framework to be built on another. In fact maybe like Flask App builder...)

I mean it's a very complex framework that we have used successfully for many years! Not wanting to slight it at all - it's amazing. But I would say it has all the components, workflows, and problems to deal with that you get with a web application framework.
